### PR TITLE
Initial dependency support

### PR DESCRIPTION
**Pull request by troibe**
**What?**
- Intitial necessary schema changes to support dependencies (`path`, `name`, `commitid`)
- Logic to download files from a variety of sources
- Saves dependency data to `~/.sc/cache/{name}-{commitid}`

**Why?**
This will allow in the next step to set file names relative to a dependency. No more searching for files or ambiguities.

**Testing:**
All the currently supported paths are tested in `test_dependency.py`

### CODE REVIEW

File: siliconcompiler/dependency.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            'path': e.g. 'ssh://github.com/xyz/xyz/'
+            'commitid': e.g. '3b94aa80506d25d5388131e9f2ecfcf4025ca866',
+            'name': 'your-dependency'
+            or
+            'path': e.g. 'https://github.com/xyz/xyz/archive/',
+            'commitid': e.g. '3b94aa80506d25d5388131e9f2ecfcf4025ca866',
+            'name': 'your-dependency'
+            or
+            'path': e.g. 'https://zeroasic.com/xyz.tar.gz',
+            'name': e.g. 'your-dependency-version-1'
```
Comment (gadfort) [inline siliconcompiler/dependency.py:45]: is this just a string now?

File: siliconcompiler/schema/schema_cfg.py [hunk 2]
```diff
@@ -2988,6 +2991,53 @@ def schema_option(cfg):
...
+            sctype='str',
+            scope='global',
+            shorthelp="Dependency name",
+            switch="-dependency_name 'dependency <str>'",
+            example=[
+                "cli: -dependency_name 'freepdk45_data freepdk45'",
+                "api: chip.set('dependency', 'freepdk45_data', 'name', 'freepdk45')"],
+            schelp="""
+            Dependency name in cache: name-commitid
+            Should be unique if no commitid specified""")
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:3036]: is this needed if the `dependency` is also a name?

File: siliconcompiler/schema/schema_cfg.py [hunk 3]
```diff
@@ -2988,6 +2991,53 @@ def schema_option(cfg):
...
+    scparam(cfg, ['dependency', dependency, 'commitid'],
+            sctype='str',
+            scope='global',
+            shorthelp="Dependency source commitid",
+            switch="-dependency_commitid 'dependency <str>'",
+            example=[
+                "cli: -dependency_commitid 'freepdk45_data 07ec4aa'",
+                "api: chip.set('dependency', 'freepdk45_data', 'commitid', '07ec4aa')"],
+            schelp="""
+            Dependency source commitid""")
```
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:3024]: maybe `version` is a bit more generic? I'm not sure about that, just a suggestion.

File: siliconcompiler/schema/schema_cfg.py (hunk 2 above)
Comment (troibe) [reply siliconcompiler/schema/schema_cfg.py:3036]: Just using the name of the dependency would also work of course.
Again might be confusing with the python package name.
For example so far I was using `siliconcompiler_data` as the python package name for the metadata but `siliconcompiler` as the name for the actual content.

File: siliconcompiler/schema/schema_cfg.py (hunk 3 above)
Comment (gadfort) [reply siliconcompiler/schema/schema_cfg.py:3024]: ref is probably good

File: siliconcompiler/dependency.py (hunk 1 above)
Comment (troibe) [reply siliconcompiler/dependency.py:45]: Yep if it just points to some random tar file on the internet.
In that case we don't have any chance to verify the `name` or `commitid`.
We still need to have a name of where we want to store the cached files.

File: siliconcompiler/schema/schema_cfg.py (hunk 2 above)
Comment (troibe) [reply siliconcompiler/schema/schema_cfg.py:3036]: Then nevermind I can just change it.

File: tests/dependency/test_dependecy.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import siliconcompiler
```
Comment (gadfort) [inline tests/dependency/test_dependecy.py:1]: rename this file to `test_dependency.py`

**Review by gadfort — CHANGES_REQUESTED** [review]
LGTM, just one minor fix

**Review by gadfort — APPROVED** [review]
LGTM

**Pull request merged**